### ISSUE TITLE
Slow Compilation for CustomIntegrators with many If Blocks

### ISSUE THREAD

**Issue by lvotapka** [openmm/openmm#2866]
Hi All,

Please forgive the very complicated script, but it's the simplest way I know how to recreate this problem.

```
from simtk.openmm.app import *
from simtk.openmm import *
from simtk import unit
from sys import stdout
import time

numsteps=4000

starttime = time.time()
prmtop = AmberPrmtopFile('dip.top')
inpcrd = AmberInpcrdFile('md-4ns.rst7')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*unit.nanometer, constraints=HBonds)

dt = 0.002*unit.picoseconds # timestep
integrator = CustomIntegrator(dt)

window1 = 0
window2 = 400
window3 = 800
window4 = 1200
window5 = 1600
window6 = 2000
window7 = 2400
window8 = 2800

interval = 100

integrator.addGlobalVariable("forceScalingFactor", 0)
integrator.addGlobalVariable("window1", window1)
integrator.addGlobalVariable("window2", window2)
integrator.addGlobalVariable("window3", window3)
integrator.addGlobalVariable("window4", window4)
integrator.addGlobalVariable("window5", window5)
integrator.addGlobalVariable("window6", window6)
integrator.addGlobalVariable("window7", window7)
integrator.addGlobalVariable("window8", window8)

integrator.addGlobalVariable("stepCount", 0)
integrator.addPerDofVariable("x0", 0);
integrator.addGlobalVariable("stepCount2", 0)
integrator.addGlobalVariable("stepCount3", 0)

integrator.beginIfBlock("stepCount >= " + str(window1))
integrator.beginIfBlock("stepCount < " + str(window2))
integrator.addComputeGlobal("forceScalingFactor", "1.0")
integrator.endBlock()
integrator.endBlock()

integrator.beginIfBlock("stepCount = " + str(window2))
integrator.addComputeGlobal("stepCount2", "0")
integrator.endBlock()

integrator.beginIfBlock("stepCount >= " + str(window2))
integrator.beginIfBlock("stepCount < " + str(window3))
integrator.addComputeGlobal("forceScalingFactor", "0.99")
integrator.addComputeGlobal("stepCount2", "stepCount2+1")
integrator.beginIfBlock("stepCount2 = " + str(interval))
integrator.addComputeGlobal("stepCount2", "0")
integrator.endBlock()
integrator.endBlock()
integrator.endBlock()
integrator.beginIfBlock("stepCount >= " + str(window3))
integrator.beginIfBlock("stepCount < " + str(window4))
integrator.addComputeGlobal("forceScalingFactor", "0.98")
integrator.endBlock()
integrator.endBlock()

integrator.beginIfBlock("stepCount = " + str(window4))
integrator.addComputeGlobal("stepCount3", "0")
integrator.endBlock()

integrator.beginIfBlock("stepCount >= " + str(window4))
integrator.beginIfBlock("stepCount < " + str(window5))
integrator.addComputeGlobal("forceScalingFactor", "0.97")
integrator.addComputeGlobal("stepCount3", "stepCount3+1")
integrator.beginIfBlock("stepCount3 = " + str(interval))
integrator.addComputeGlobal("stepCount3", "0")
integrator.endBlock()
integrator.endBlock()
integrator.endBlock()


integrator.beginIfBlock("stepCount >= " + str(window5))
integrator.beginIfBlock("stepCount < " + str(window6))
integrator.addComputeGlobal("forceScalingFactor", "0.96")
integrator.endBlock()
integrator.endBlock()
'''
integrator.beginIfBlock("stepCount >= " + str(window6))
integrator.beginIfBlock("stepCount < " + str(window7))
integrator.addComputeGlobal("forceScalingFactor", "0.95")
integrator.endBlock()
integrator.endBlock()

integrator.beginIfBlock("stepCount >= " + str(window7))
integrator.beginIfBlock("stepCount < " + str(window8))
integrator.addComputeGlobal("forceScalingFactor", "0.94")
integrator.endBlock()
integrator.endBlock()
'''
integrator.addUpdateContextState();
integrator.addComputePerDof("x0", "x");
integrator.addComputePerDof("v", "v+dt*f/m");
integrator.addComputePerDof("x", "x+dt*v");
integrator.addConstrainPositions();
integrator.addComputePerDof("v", "(x-x0)/dt");

integrator.addComputeGlobal("stepCount", "stepCount+1")

platform = Platform.getPlatformByName('CUDA')
properties = {'CudaDeviceIndex': '1', 'CudaPrecision': 'mixed'}
simulation = Simulation(prmtop.topology, system, integrator, platform, properties)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.context.setVelocitiesToTemperature(300*unit.kelvin)

simulation.reporters.append(StateDataReporter(stdout, interval, step=True, potentialEnergy=True, temperature=True))

print("time to prepare:", time.time() - starttime)

simulation.step(interval)
```

Now, the way the script is, it will run without taking too much time, but if you uncomment the window6 to window7 If Block, you'll notice a remarkable increase in time to start the simulation. If you uncomment both If Blocks, then you'll notice quite a bit more time taken before the simulation starts.

Here are the zipped input files:
[dip.top.zip](https://github.com/openmm/openmm/files/5285062/dip.top.zip)
[md-4ns.rst7.zip](https://github.com/openmm/openmm/files/5285063/md-4ns.rst7.zip)

I know that this is an unusually large number of if blocks, but it would be difficult for me to accomplish the task I'm trying to without this many if blocks. Is this a problem that can be resolved within the code, or do I just have no choice but to use less if blocks?

**Comment by peastman** [comment]
Unfortunately, the startup time scales exponentially in the number of `if` or `while` blocks.  This comment in the code explains exactly what it's doing and why.

https://github.com/openmm/openmm/blob/94d7225bfe59e2ce80fc4edfc1828d15d73aa1f5/openmmapi/src/CustomIntegratorUtilities.cpp#L191-L207

I'm not sure I entirely follow the logic of your code.  Is it just decreasing `forceScalingFactor` by 0.1 every 400 steps?  You could do that with just a single block:

```python
integrator.beginIfBlock("stepCount = 400")
integrator.addComputeGlobal("forceScalingFactor", "forceScalingFactor-0.01")
integrator.addComputeGlobal("stepCount", "0")
integrator.endBlock()
```

Or are you doing something a bit more complicated than that?  You also have the `stepCount2` and `stepCount3` variables, but I can't see that they ever affect anything.

**Comment by lvotapka** [comment]
First off, thanks for getting back to me about this, @peastman. 

The script I posted here is only intended to illustrate the observation that many if blocks seem to increase the startup time - it serves no other useful purpose. I am fully aware that I can accomplish a gradual decrease in force scaling much more easily using the technique you suggested. The integrator that I am really trying to design needs many if blocks, but was too complex to post here on a Github issue.

And now I understand better from your comment in the code why there is an exponential time increase for the compilation with many if blocks. With that in mind, I'll see if there are any opportunities to cut down on the number of if blocks.

If you have any other general suggestions for how to decrease compilation time, I would welcome the opportunity to hear them.

**Comment by peastman** [comment]
This issue is addressed in #4191.  When doing the analysis of flow control described above, it ignores blocks that neither invalidate nor use forces or energy.  With this change, the script above takes only two seconds, even with no blocks commented out.

I hesitate to say this completely fixes the general problem.  The cost of the analysis still scales exponentially in the number of *significant* blocks (ones that use or invalidate forces or energy).  There could still be integrators for which it's very expensive.  But for the particular script in this issue it completely eliminates the problem, and it hopefully will do so for many other integrators too.
